### PR TITLE
Require all fields for addSubmission function

### LINKED ISSUE #261: fix: createSubmissions Service Functions
# Summary
Currently, a user is able to create a submission without including a `googleMapsUrl`, `healthcareProfessionalName`, or `spokenLanguages`. We want to prevent users from doing this.

# Requirements
- [ ] Add logic to require users to include all fields.

### PR DESCRIPTION
**Pull request by evan-desu**
Previously, submissions with empty fields were accepted and added to the database. This ensures that all required fields (googleMapsUrl, healthcareProfessionalName, and spokenLanguages) are provided.

Resolves #261 

# What changed
1. Added validation to ensure` googleMapsUrl` and `healthcareProfessionalName` are not empty.
2. Validated `spokenLanguages` and ensured it's not an empty array.

# Testing instructions
Copy and paste this into the GraphQL Studio:

**Operation:**
```
mutation Mutation($input: SubmissionInput) {
  createSubmission(input: $input) {
    id
      }
}
```

**Variables:**
```
{
  "input": {
    "googleMapsUrl": " ",
    "healthcareProfessionalName": "Dr Bob",
    "spokenLanguages": [
      {
        "iso639_3": "eng",
        "nameJa": "English",
        "nameEn": "English",
        "nameNative": "English"
      }
    ]
  }
}
```
You should receive an error similar to this:
```
{
  "data": {
    "createSubmission": null
  },
  "errors": [
    {
      "message": "Error adding submission: Error: Fields cannot be empty strings.",
      "locations": [
        {
          "line": 3,
          "column": 3
        }
      ],
```

### CODE REVIEW

File: src/services/submissionService.ts [hunk 1]
```diff
@@ -118,7 +118,23 @@ function convertToDbSubmission(submission: gqlType.SubmissionInput):
 export const addSubmission = async (submissionInput: gqlType.SubmissionInput): 
     Promise<dbSchema.Submission> => {
     try {
-        const dbSubmission = convertToDbSubmission(submissionInput)
+        if (!submissionInput.googleMapsUrl.trim() || !submissionInput.healthcareProfessionalName.trim()) {
```
Comment (ermish) [inline src/services/submissionService.ts:121]: Let's move all the validation work into a dedicated validation function

Comment (ermish) [reply src/services/submissionService.ts:121]: Should check for null/undefined before calling trim

File: src/services/submissionService.ts [hunk 2]
```diff
@@ -118,7 +118,23 @@ function convertToDbSubmission(submission: gqlType.SubmissionInput):
...
+            throw new Error('Submission fields cannot be empty strings.')
+        }
+
+        const nonNullLanguages = submissionInput.spokenLanguages
+            .filter(lang => lang !== null) as gqlType.SpokenLanguageInput[]
```
Comment (ermish) [inline src/services/submissionService.ts:126]: Null is actually different then undefined. We should check for both. 
You can likely use `!!lang` for the check

File: src/services/submissionService.ts [hunk 3]
```diff
@@ -118,7 +118,23 @@ function convertToDbSubmission(submission: gqlType.SubmissionInput):
...
+
+        const validatedLanguages = mapAndValidateSpokenLanguages(nonNullLanguages)
+
+        if (!validatedLanguages || validatedLanguages.length === 0) {
```
Comment (ermish) [inline src/services/submissionService.ts:130]: Move this into the map and validate function. 
This is probably a good time for us to introduce a proper errors object. 
Have it return a `hasErrors` Boolean and and errors array where each error is an object with a `message` field and `property` field
Example

```
result: {
    hasErrors: true,
    errors: [
        { field: "spokenlanguages[1].type",
          message: "the type can only be english or Japanese" }]
}
```

**Review by ermish — CHANGES_REQUESTED** [review]
We can talk on slack if you need help with some of this. 
It's a new design

File: src/services/submissionService.ts (hunk 3 above)
Comment (ermish) [reply src/services/submissionService.ts:130]: update: 
1) the error message should be based on the custom errors type.
2) we should make this result object a typscript type that can be used across all the services.

Comment (ermish) [reply src/services/submissionService.ts:130]: update #2: I have a PR with a results framework set up that should make this a bit easier and more consistent.

**Review by theyokohamalife — APPROVED** [review]
LGTM!

File: .env.prod [hunk 4]
```diff
@@ -7,6 +7,4 @@ FIRESTORE_PROJECT_ID = "find-a-doc-japan"
 # https://console.firebase.google.com/u/0/project/find-a-doc-japan/overview
 # DO NOT COMMIT THESE VALUES TO GIT!!!
 
-# FIRESTORE_SERVICE_ACCOUNT = 'enteryourkeyhere'
-
-
+# FIRESTORE_SERVICE_ACCOUNT = 'enteryourkeyhere'
```
Comment (theyokohamalife) [inline .env.prod:10]: I think you can leave this out since it's already on main.

Comment (evan-desu) [reply .env.prod:10]: @theyokohamalife I'll be sure to remove it (or figure out how to do that).

File: src/services/submissionService.ts [hunk 5]
```diff
@@ -158,6 +168,16 @@ export const updateSubmission = async (id: string, updatedFields: Partial<dbSche
     }
 }
 
+const validateSubmissionInputFields = (input: gqlType.SubmissionInput) => {
+    if (
+        (input.googleMapsUrl === null || input.googleMapsUrl === undefined || !input.googleMapsUrl.trim()) ||
```
Comment (ermish) [inline src/services/submissionService.ts:173]:
```suggestion
        (input.googleMapsUrl || !input.googleMapsUrl.trim()) ||
```

didn't see this sooner. you can simplify a null and undefined check with a basic truthy check. (truthy checks for both and empty strings, but you need to be careful if things like 0 are valid options.

File: src/services/submissionService.ts [hunk 6]
```diff
@@ -158,6 +168,16 @@ export const updateSubmission = async (id: string, updatedFields: Partial<dbSche
...
+        (input.healthcareProfessionalName === null || input.healthcareProfessionalName === undefined ||
```
Comment (ermish) [inline src/services/submissionService.ts:174]:
```suggestion
        (input.healthcareProfessionalName || 
```

File: src/services/submissionService.ts [hunk 7]
```diff
@@ -196,7 +216,9 @@ function gqlSpokenLanguageToDbSpokenLanguage(lang: gqlType.SpokenLanguage):
 
 export const mapAndValidateSpokenLanguages = (spokenLanguages: gqlType.SpokenLanguageInput[]): 
     dbSchema.SpokenLanguage[] | undefined => {
-    if (!spokenLanguages || ! spokenLanguages.length) { return undefined }
+    if (!spokenLanguages || spokenLanguages.length === 0) {
```
Comment (ermish) [inline src/services/submissionService.ts:219]:
```suggestion
    if (!spokenLanguages) {
```

File: src/resolvers.ts [hunk 8]
```diff
@@ -133,7 +133,7 @@ const resolvers = {
         }) => {
             try {
                 for (const value of Object.values(args.input)) {
-                    if (!value) {
```
Comment (ermish) [inline src/resolvers.ts:136]: Wait. This should be !value. It's the equivalent.

File: src/resolvers.ts [hunk 9]
```diff
@@ -157,8 +157,11 @@ const resolvers = {
                     updatedDate: new Date().toISOString()
                 }
                 const newSubmission = await submissionService.addSubmission(submissionData)
-    
-                return newSubmission
+
+                if (newSubmission.hasErrors) {
+                    throw new Error('Failed to create submission.')
+                }
+                return newSubmission.data
```
Comment (ermish) [inline src/resolvers.ts:164]: Perfect!

File: src/resolvers.ts [hunk 10]
```diff
@@ -157,8 +157,11 @@ const resolvers = {
                     updatedDate: new Date().toISOString()
                 }
                 const newSubmission = await submissionService.addSubmission(submissionData)
```
Comment (ermish) [inline src/resolvers.ts:159]: nit: addSubmissionResult would be a more accurate name.

File: src/resolvers.ts [hunk 11]
```diff
@@ -177,9 +180,17 @@ const resolvers = {
...
+                
+                if (args.input.spokenLanguages) {
+                    const spokenLanguagesResult = submissionService
+                        .mapAndValidateSpokenLanguages(args.input.spokenLanguages)
+                    
+                    if (spokenLanguagesResult.hasErrors) {
+                        throw new Error('Invalid spoken Language provided.')
+                    }
+                    updatedSpokenLanguages = spokenLanguagesResult.data
+                }
```
Comment (ermish) [inline src/resolvers.ts:193]: This is ok for now. 
@evan-desu  can you create a new issue to move all of the mapping and validation work into the service? 
The idea is to keep your resolvers as simple as possible and have the service method do all the work.

File: src/services/submissionService.ts [hunk 12]
```diff
@@ -158,6 +183,43 @@ export const updateSubmission = async (id: string, updatedFields: Partial<dbSche
     }
 }
 
+const validateSubmissionInputFields = (input: gqlType.SubmissionInput): Result<dbSchema.Submission> => {
+    const validatedSubmissionResult: Result<dbSchema.Submission> = {
+        hasErrors: false,
+        errors: []
+    }
+    
+    if (input.googleMapsUrl === null || input.googleMapsUrl === undefined || !input.googleMapsUrl.trim()) {
```
Comment (ermish) [inline src/services/submissionService.ts:192]: Use shorthand 😎

```suggestion
    if (!input.googleMapsUrl || !input.googleMapsUrl.trim()) {
```

File: src/services/submissionService.ts [hunk 13]
```diff
@@ -158,6 +183,43 @@ export const updateSubmission = async (id: string, updatedFields: Partial<dbSche
...
+        validatedSubmissionResult.hasErrors = true
+        validatedSubmissionResult.errors?.push({
+            field: 'googleMapsUrl',
+            errorCode: ErrorCode.MISSING_INPUT,
+            httpStatus: 400
+        })    
+    }
+
+    if (input.healthcareProfessionalName === null || input.healthcareProfessionalName === undefined ||
```
Comment (ermish) [inline src/services/submissionService.ts:201]:
```suggestion
    if (!input.healthcareProfessionalName || 
```

File: src/services/submissionService.ts [hunk 14]
```diff
@@ -158,6 +183,43 @@ export const updateSubmission = async (id: string, updatedFields: Partial<dbSche
...
+        validatedSubmissionResult.hasErrors = true
+        validatedSubmissionResult.errors?.push({
+            field: 'googleMapsUrl',
+            errorCode: ErrorCode.MISSING_INPUT,
+            httpStatus: 400
+        })    
+    }
+
+    if (input.healthcareProfessionalName === null || input.healthcareProfessionalName === undefined || 
+        !input.healthcareProfessionalName.trim()) {
```
Comment (ermish) [inline src/services/submissionService.ts:202]: We should put a limit on all these string lengths. 256 should be more than enough

Comment (ermish) [reply src/services/submissionService.ts:202]: N/m ha just saw the next one. 

We should probably limit the map url to 1028

**Review by ermish — APPROVED** [review]
Great! Logic looks sound. Just some small optimizations and we're good to go

File: .env.prod (hunk 4 above)
Comment (theyokohamalife) [reply .env.prod:10]: @evan-desu  You should be able to run `git checkout main .env.prod` and then commit and push to have this removed.

Comment (evan-desu) [reply .env.prod:10]: @theyokohamalife I'll try to do that later this evening. I gave up yesterday 😅

**Review by ermish — APPROVED** [review]
Ship it! :shipit:

**Pull request merged**